hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java

```diff
@@ -61,6 +69,63 @@ public void remove() {
         };
     }
 
+    /**
+     * Lazily filters iterated elements.
+     *
+     * @return a new iterable object which
+     * has an iterator capable of filtering elements
+     */
+    public static <T> Iterable<T> filter(Iterable<T> iterable, Predicate<T> filter) {
+        Iterator<T> givenIterator = iterable.iterator();
+
+        Iterator<T> filteringIterator = new Iterator<T>() {
+            private T next;
+
+            @Override
+            public boolean hasNext() {
+                boolean hasNext = false;
+                while (givenIterator.hasNext()) {
+                    T temp = givenIterator.next();
+                    if (filter.test(temp)) {
+                        next = temp;
+                        hasNext = true;
+                        break;
+                    }
+                }
+                return hasNext;
+            }
+
+            @Override
+            public T next() {
+                if (next == null) {
```

Comment:
We should `throw NoSuchElementException` only if there is no next element not expecting that the user called `hasNext()` before that. I'd put it like
if (!hasNext()) {
    throw new NoSuchElementException()
}

Reply:
fixed.